common/interfaces/card_interface/nfc.c

```diff
@@ -303,29 +303,35 @@ ISO7816 nfc_unpair() {
   return status_word;
 }
 
-ISO7816 nfc_list_all_wallet(uint8_t recv_apdu[], uint16_t *recv_len) {
-  ASSERT(recv_apdu != NULL);
-  ASSERT(recv_len != NULL);
+ISO7816 nfc_list_all_wallet(wallet_list_t *wallet_list) {
+  // Select card before.
+  ISO7816 status_word;
+  uint8_t send_apdu[300], send_len;
 
-  // Select card before. recv_len receives the length of response APDU. It also
+  // Use the same buffer to receive NFC apdu
+  uint8_t *recv_apdu = send_apdu;
+
+  // recv_len receives the length of response APDU. It also
   // acts as expected length of response APDU.
-  ISO7816 status_word;
-  uint8_t send_apdu[255], send_len;
-  *recv_len = 236;
+  uint16_t recv_len = 236;
+
   send_len = create_apdu_list_wallet(send_apdu);
 
   nfc_secure_comm = true;
   uint32_t system_clock = uwTick;
   ret_code_t err_code =
-      nfc_exchange_apdu(send_apdu, send_len, recv_apdu, recv_len);
+      nfc_exchange_apdu(send_apdu, send_len, recv_apdu, &recv_len);
   LOG_SWV("List wallet in %lums\n", uwTick - system_clock);
 
   if (err_code != STM_SUCCESS) {
     return err_code;
   } else {
-    status_word = (recv_apdu[*recv_len - 2] * 256);
-    status_word += recv_apdu[*recv_len - 1];
+    status_word = (recv_apdu[recv_len - 2] * 256);
+    status_word += recv_apdu[recv_len - 1];
 
+    if (status_word == SW_NO_ERROR) {
+      apdu_extract_wallet_list(wallet_list, recv_apdu, recv_len);
+    }
     return status_word;
   }
   return 0;
```

Comment:
I think the default return code should be err_code instead of 0. Can modify the if condition accordingly.

Reply:
This logic was not in scope for this pull request and therefore I did not consider it.

On quick review, it seems that control will never reach this statement here, so we should avoid changing without background info